### PR TITLE
Fix cloning entities when using lazy-ghost proxies

### LINKED ISSUE #10817: Wrong LazyGhost initialization on cloned entity
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.15.3

#### Summary

When you clone an entity with a oneToOne relationship, the cloned entity is filled with data that already exists and has nothing to do with it.

#### Current behavior

The clone is reloaded from the database even though it is a new entity

#### How to reproduce

You can find a reproducer here: https://github.com/Erwui/lazy_ghost_bug
Go on "/"

#### Expected behavior

When setting "enable_lazy_ghost_objects" to false the clone is correct

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Fixes #10817
Replaces #10818
Replaces https://github.com/symfony/symfony/pull/50899

### CODE REVIEW

File: lib/Doctrine/ORM/Proxy/ProxyFactory.php [hunk 1]
```diff
@@ -98,6 +86,9 @@ public function __serialize(): array
      */
     private $identifierFlattener;
 
+    /** @var ProxyDefinition[] */
```
Comment (stof) [inline lib/Doctrine/ORM/Proxy/ProxyFactory.php:89]: this looks like `array<class-string, ProxyDefinition>`

Comment (nicolas-grekas) [reply lib/Doctrine/ORM/Proxy/ProxyFactory.php:89]: I borrowed from the parent class ;)

**Comment by nicolas-grekas** [comment]
All green :green_apple:

**Review by stof — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @nicolas-grekas !

**Comment by mpdude** [comment]
Can you guys provide an example when/how it is useful to clone entities and what the semantics of this operation are?

Does the new (cloned) entity let go of its ID? Should the EntityManager know about it? Is it in `MANAGED` or `DETACHED` state etc.?

**Comment by stof** [comment]
@mpdude the cloned object will not be managed by the entity manager

**Comment by mpdude** [comment]
When the entity that is cloned is an uninitialized proxy at that point in time, what happens when I call a method on the clone that triggers loading?

**Comment by mpdude** [comment]
Coming from #3037 and #10785 where one source of "weird" problems is corruption of internal UoW state-tracking fields. This may be caused when using proxies directly created by the ProxyFactory. 

My concern is that cloning a proxy object might be no different:

Suddenly there are two objects claiming to proxy the same entity. But, "there can be only one", only the proxy object the UoW knows about should be initialized and is the sole object representing that entity.

**Comment by TheBoringGuy1908** [comment]
> When the entity that is cloned is an uninitialized proxy at that point in time, what happens when I call a method on the clone that triggers loading?

I stumbled upon this while investigating an issue that seems to be related to this fix, and your suspicion seems to be confirmed.

Since doctrine/orm 2.15.4 I encounter the following problem:
- I have a lazy-loaded entity `original`
- I clone _original_ into `clone`
- I read some data from `original`, causing it to load the remaining data
- I modify some data on `original`

Up until now, everything is as expected. However, once I read some data from `clone` , it reads the missing data back from the database and populates it into `clone` **and** `original`, causing the modifications on `original` to be lost.

**Comment by nicolas-grekas** [comment]
Please open a dedicated issue with a reproducer. Commenting on closed issues/PRs isn't tracked.

**Pull request merged**